lib/useButter.js

```diff
@@ -0,0 +1,225 @@
+            "Content-Type": "application/json",
+            "X-Butter-Client": `JS/${ BUTTER_CLIENT_VERSION }`
+        });
+        
+        // Append additional header for server-side environments
+        if (typeof window === "undefined") {
+            butterHeaders.append("Accept-Encoding", "gzip");
+        }
+        
+        // Prepare the request configuration using the onRequest hook
+        const {
+            config,
+            headers,
+            params
+        } = await useOnRequest(
+            apiEndpoint,
+            auth_token,
+            {
+                cancelRequest,
+                config: userConfig,
+                headers: butterHeaders,
+                params: localParams,
+                type,
+            }
+        );
+        
+        try {
+            // Set up fetch timeout and signal
+            const {
+                cleanup,
+                signal
+            } = useFetchTimeout(fetchController, friendlyType, config.timeout);
+
+            // Perform the API request
+            const response = await fetch(
+                `${ apiEndpoint }?${ new URLSearchParams(params) }`,
+                {
+                    cache: config.cache,
+                    method: "GET",
+                    headers,
+                    signal
+                }
+            );
+            
+            // Clean up after the fetch timeout
+            cleanup();
+            
+            // Process the response using the onResponse hook
+            const {
+                data,
+                meta,
+                ...errors
+            } = await useOnResponse(
+                response,
+                {
+                    config,
+                    // pass the request as user won't see our
+                    // changes for auth_token or testMode
+                    // remove testMode params and auth_token in onResponse
+                    params,
+                    type
+                }
+            );
+            
+            // Check for errors in the response and throw if present
+            if (Object.keys(errors).length) {
+                throw { errors, config, params };
+            }
+
+            // Return the successful response data and metadata
+            return { data, meta, ...errors };
+        } catch (errorPayload) {
+            // Handle errors caught from the API request
+            if (errorPayload.config) {
+                // Process errors using the onError hook with the attached config
+                useOnError(
+                    errorPayload.errors,
+                    {
+                        config: errorPayload.config,
+                        params: errorPayload.params,
+                        type
+                    }
+                );
+                
+                // Return null data and the errors
+                return { data: null, errors: errorPayload.errors };
+            } else {
```

Comment:
We do `return` in `if` part, so we could skip `else` part to make it less nested. WDYT?

Reply:
Different type of errors. The first `if` statement handles data errors sent back by ButterAPI, the `else` handles browser/fetch/404/500 errors. And in the `else` statement we also look to see if it was a Fetch abort and look for the reasoning. I think this leaves the `if else` statement more readable. Otherwise we'd be left with a more complex if statement looking to see what sort of error we got (Butter server response, browser/404/500, or fetch abort)